### ISSUE TITLE
Motor starts and stops with Pixhawk 1.6 ESC with constant setpoint

### ISSUE THREAD

**Issue by CarlOlsson** [thiemar/vectorcontrol#12]
The ESC communicates with the GUI but when the setpoint is set to something higher than approx. 30 the motor starts and stops. @bendyer From the plots, could you indicate what is wrong?

![11](https://cloud.githubusercontent.com/assets/7795133/11278203/2f1c4078-8eea-11e5-98df-8162b9f4de0c.png)

![12](https://cloud.githubusercontent.com/assets/7795133/11278208/32a0dfe2-8eea-11e5-9124-97b7b7df3af4.png)

![13](https://cloud.githubusercontent.com/assets/7795133/11278209/34f8bed6-8eea-11e5-8735-463829af4d62.png)

- Erased the microcontroller with a JLink Debugger
- Flashed the current master bootloader to the ESC with a JLink Debugger
- Flashed the current master fw to the ESC via the Pixhawk

We built the fw using gcc 4.8.3, the git hash is 544c8 and the binaries are in the ESC_bringup Drive.

**Comment by bendyer** [comment]
Are you still using the 0.021 Ω motor? If so, the current sense measurements at the very start of spin-up look about right; it's getting around 5 A from an applied voltage lower than 0.2 V, so it's definitely well under the 0.1 Ω it estimated during startup.

Current offsets are also minimal, so overall the charts are looking pretty good, but the motor is stopping because the estimator loses track of the rotor position (you can see it starting to slip when you get the noise in the current chart).

For reference, on the "Current draw" chart, red is the current setpoint (i.e. the target torque value), blue is the D current, and black is the Q current. The inner control loop tries to keep the D current at zero, and the Q current at the setpoint; when you see oscillation around those points, either you have a lot of current sense noise, or the estimator is losing sync.

The output of the inner control loop is shown on the "Voltage" chart; the D and Q voltages are blue and black respectively.

**Comment by CarlOlsson** [comment]
Yes we are still using the 0.021 Ω motor. Thanks for the explanation of the labels etc.

Looking at the getting started guide we will try again with adjusting the gain, or do you have any other input on how to proceed? What kind of logs would be most interesting for figuring out if it is an HW, SW or configuration issue?

**Comment by bendyer** [comment]
This is a measurement problem, so I don't think any parameter changes will be useful. I expect it's related to current shunt sample timing during the resistance and inductance measurement, since that was the hardest part to get right during the porting process; it's entirely possible it's not as robust as it needs to be for the Pixhawk ESC 1.6 hardware.

**Comment by bendyer** [comment]
In 68e34de I have introduced per-board motor parameter measurement configuration, and adjusted the PX4ESC's configuration to improve its measurement performance. I did not see the specific issue you had with the motors I'm using, but did see similar issues which arose from poor measurement accuracy.

The new parameters increase measurement currents quite significantly in order to overcome current shunt measurement noise, so you should see better results.

One other point is you may need to reduce the gain to 0.5–0.75 and the braking to 0.2–0.5 to improve stability during significant step changes in throttle.

**Comment by CarlOlsson** [comment]
Hi Ben, thanks for the update! Unfortunately we still have the same problem, see plots below:

The following plots are for constant setpoint (i.e. we do not start and stop intentionally) and different parameter configurations:
![constant1](https://cloud.githubusercontent.com/assets/7795133/11324439/2e4c6016-9132-11e5-8b20-434f6df97bc2.png)

![constant2](https://cloud.githubusercontent.com/assets/7795133/11324440/31bc4496-9132-11e5-99ca-f76a47aaefc8.png)

![constant3](https://cloud.githubusercontent.com/assets/7795133/11324442/3488729e-9132-11e5-9d8b-70956aa73531.png)

If we set the setpoint to 10 the motor does not stop. Here is a plot from such a case:
![working_setpoint_10](https://cloud.githubusercontent.com/assets/7795133/11324449/597f1d46-9132-11e5-96e8-949df6719b14.png)

**Comment by bendyer** [comment]
Thanks @CarlOlsson, those charts provide the information I need to continue investigating.

**Comment by CarlOlsson** [comment]
No problem! Just ping us if you want any other logs etc!

**Comment by bendyer** [comment]
I have been testing using a Scorpion S3026-II 790 Kv, which has reasonably similar specs to the Hacker A40-14S V2 you've been using.

I was able to reproduce the measurement issues you've seen, as well as the starting/stopping behaviour, which is due to a loss of sync during the sudden change in target voltage when it comes out of the initial speed ramp. 

Here are some changes which significantly improve performance in my tests:
https://github.com/thiemar/vectorcontrol/tree/pixhawk_esc_bringup

Please try them out and let me know how you go. My suggested parameters are:

Current limit: 12 A
Voltage limit: 12 V (for now)
Acceleration limit: 0.25 V
Bandwidth: 50 Hz
Gain: 1.0
Braking: 0.2

**Comment by CarlOlsson** [comment]
**Preparation:**
- Erased flash on the Pixhawk ESC 1.6
- Flashed the bootloader
- Flashed the fw built from the pixhawk_esc_bringup with hash 6dd12173
- Nice sound on startup!
- Changed parameters to the ones suggested.

**Results when running the motor with different setpoints:**
- The Motor does not start and stop anymore
- Sometimes (rarely) the motor does not start at all, we then stopped and started and the motor starts to spin
- **The setpoint value did not seam to influence the rpm much, at least not above a setpoint of 20**
- Even with a setpoint of 100 (or even 300) the current draw was not more than 0.8 A

**Some plots:**
Setpoint 10:
![setpoint_10](https://cloud.githubusercontent.com/assets/7795133/11374019/4fdd2f32-92d4-11e5-9bb8-b0792e65d37b.png)

Setpoint 20:
![setpoint_20](https://cloud.githubusercontent.com/assets/7795133/11374030/5590ec48-92d4-11e5-8640-52f9549f0a80.png)

Setpoint 30:
![setpoint_30](https://cloud.githubusercontent.com/assets/7795133/11374035/59d6b300-92d4-11e5-8fe2-08de2166327e.png)

Setpoint 50:
![setpoint_50](https://cloud.githubusercontent.com/assets/7795133/11374037/5d6c4052-92d4-11e5-88c6-e3d593173a14.png)

Setpoint 70:
![setpoint_70](https://cloud.githubusercontent.com/assets/7795133/11374059/77f77996-92d4-11e5-9aac-c5dc12c1395d.png)

Setpoint 100:
![setpoint_100](https://cloud.githubusercontent.com/assets/7795133/11374065/8332f89e-92d4-11e5-8cf2-cf0272bfc09f.png)

Setpoint 300:
![selection_300](https://cloud.githubusercontent.com/assets/7795133/11374072/8b3536f6-92d4-11e5-976a-394397bdbdd6.png)

We tried increasing the maximum current and voltage but with the same qualitative behaviour

**Comment by mhkabir** [comment]
@bendyer @CarlOlsson 
I'm trying to bring up a Pixhawk ESC 1.6 with some Turnigy motors. (http://www.hobbyking.com/mobile/viewproduct.asp?idproduct=38456) 

I'm facing the exact same issues. I'm on the ```pixhawk_esc_bringup``` branch.
